src/identity/processor.rs

```diff
@@ -20,47 +20,50 @@ use crate::database::query::DatabaseQuery;
 use crate::database::types::{BatchEntry, BatchType};
 use crate::database::Database;
 use crate::ethereum::{Ethereum, ReadProvider};
-use crate::identity_tree::{Canonical, Intermediate, TreeVersion, TreeWithNextVersion};
+use crate::identity_tree::{Canonical, Hash, Intermediate, TreeVersion, TreeWithNextVersion};
+use crate::prover::identity::Identity;
+use crate::prover::repository::ProverRepository;
 use crate::prover::Prover;
 use crate::utils::index_packing::pack_indices;
 use crate::utils::retry_tx;
 
 pub type TransactionId = String;
 
 #[async_trait]
-pub trait IdentityTransactionManager: Send + Sync + 'static {
+pub trait IdentityProcessor: Send + Sync + 'static {
     async fn commit_identities(&self, batch: &BatchEntry) -> anyhow::Result<TransactionId>;
-    async fn produce_transaction_finalizer(
-        &self,
-    ) -> anyhow::Result<Box<dyn IdentityTransactionFinalizer>>;
+    async fn produce_transaction_finalizer(&self) -> anyhow::Result<Box<dyn IdentityFinalizer>>;
 
     async fn await_clean_slate(&self) -> anyhow::Result<()>;
 
     async fn mine_transaction(&self, transaction_id: TransactionId) -> anyhow::Result<bool>;
+
+    async fn tree_init_correction(&self, initial_root_hash: &Hash) -> anyhow::Result<()>;
 }
 
 #[async_trait]
-pub trait IdentityTransactionFinalizer: Send + Sync + 'static {
+pub trait IdentityFinalizer: Send + Sync + 'static {
     async fn finalize_identities(
         &mut self,
         processed_tree: &TreeVersion<Intermediate>,
         mined_tree: &TreeVersion<Canonical>,
     ) -> anyhow::Result<()>;
 }
```

Comment:
Is there a reason that we are creating a trait with corresponding `OffchainFinalizer` and `OnchainFinalizer` structs rather than just implementing a `finalize_roots()` method on the `IdentityProcessor` trait?

With this approach, `identity_processor.finalize_roots()` could return a future or spawned task and we could simply invoke this in the task manager.

Reply:
I needed a moment to recall the reason. :D Only `IdentityFinalizer `is mutable while `IdentityProcessor` is not (as stored in app object). Mutability is required by blockchain scanners (like here https://github.com/worldcoin/signup-sequencer/pull/743/files/665a3574a4305633208e1b80eff636ed2f006245#diff-0fa352a4471262fa3eb265e56414ea649dd662b9abde25e8d0e382dd256417baR381). I could go with https://doc.rust-lang.org/book/ch15-05-interior-mutability.html but I personally don't like to use "unsafe" parts. I prefer to let compiler still help with the work. When `IdentityFinalizer` is returned from `IdentityProcessor` then only there the variable (https://github.com/worldcoin/signup-sequencer/pull/743/files/665a3574a4305633208e1b80eff636ed2f006245#diff-5535a609ff4d5028c3f7b86e1fd69d3dd6c5ceeb6bb09938f52a02e8fe55234bR6) is mutable and mutability is "bind" to the owner (here it is `finalize_identities` task).
If you think there is a better solution I'm opened to any other implementation here. :)

Reply:
>  I personally don't like to use "unsafe" parts
Interior mutability via `Refcell` or `Mutex` itself not unsafe, even though there is unsafe logic within the implementation. The compiler will still help you if you use a`Mutex` or similar for the mutable components. With that being said, you would still need to make sure you are not acquiring the lock for longer than necessary. 

I think it might be a little cleaner to have the finalization logic directly in the `IdentityProcessor` but I don't have a strong opinion and would defer to @Dzejkop if there is a preference.

Reply:
Now that I'm thinking about it we should move `BlockScanner` into `IdentityProcessor` (or somewhere else).

If we create it everytime `finalize_identities` restarts we risk missing blocks that we're not scanning

Reply:
And in that case it makes more sense to put it behind a `Mutex`

Reply:
Done. Thanks for suggestion! Looks much better now. :)